base/abstractarraymath.jl

```diff
@@ -4,6 +4,15 @@
 
 isreal(x::AbstractArray) = all(isreal,x)
 iszero(x::AbstractArray) = all(iszero,x)
+isone(x::AbstractVector) = false # vectors are not multiplicative identities
+function isone(x::AbstractMatrix)
+    m, n = size(x)
+    m != n && return false # only square matrices can satisfy x == one(x)
+    @inbounds for i in 1:m
```

Comment:
I tried a loop over all elements, but it was about 5-6 times slower than just checking for equality with the identity matrix. The allocation, however, has a massive overhead, meaning that every `isone(A)` basically takes the same amount of time, independent of whether or not it is the identity matrix. Checking the diagonal will very quickly find most matrices that aren't the identity, unless they happen to be one all the way, and then have the penalty of the allocation overhead.